src/utp_stream.cpp

```diff
@@ -281,7 +281,7 @@ struct utp_socket_impl
 		, m_written(0)
 		, m_receive_buffer_size(0)
 		, m_read_buffer_size(0)
-		, m_in_buf_size(1024 * 1024)
+		, m_in_buf_size(8 * 1024 * 1024)
```

Comment:
the problem of not connecting this back to the socket option (which I suppose is broken right now, is that right?) is that this is the mechanism used to provide back-pressure to senders, when the next layer isn't consuming data quick enough. Specifically, when the user sets a download rate limit, libtorrent won't read from the socket as fast as it can, but only as fast as it receives download quota. In that situation, you can end up allocating a lot of memory for these buffers.

I know, the download rate limiter should limit the bittorrent requests instead, and still read as fast as possible. It's on my list, but the list is pretty long.

Reply:
As I understand, there are 2 levels of buffering of incoming packets (not counting the user-provided `m_read_buffer`):
`[OS-managed UDP socket buffer] -> [utp_stream_impl receive buffer]`.

`settings_pack::recv_socket_buffer_size` sets the maximum size of the OS-managed UDP socket buffer. The maximum size for the `utp_socket_impl.m_receive_buffer` is stored in `utp_socket_impl.m_in_buf_size`, and it is only set once at initialization of `utp_socket_impl`. However, the name `utp_socket_impl.m_receive_buffer_size` denotes:
```
    // the sum of all packets stored in m_receive_buffer
    boost::int32_t m_receive_buffer_size;
```
This creates ambiguity in naming.

| Level        | OS socket                                | utp_socket_impl                       |
|--------------|------------------------------------------|-----------------------------------------|
| Current size | ???                                      | `m_receive_buffer_size` |
| Maximum size | `recv_socket_buffer_size` | `m_in_buf_size`         |

If the current implementation of download rate control in libtorrent results in `m_receive_buffer` always filled up to `m_in_buf_size` limit, we really want to set it as low as possible to keep the memory usage reasonable. But we need to provide at least some way to increase it for high-latency connections.

To do this, we could take one of the following approaches:
1. Bound it to the OS socket size, and increase it proportionally if user changes `recv_socket_buffer_size`.
2. Create a new option in `settings_pack`.
3. Make it scale automatically, dependent on latency.

Which one of these you find the best for libtorrent?

Reply:
I think (1) is the simplest and safest option. I'm hesitant to add another option, as there are so many already, and I would be a bit worried about (3) because there would likely need to be a lot of heuristics and instability in practice. Latency isn't just caused by long distances, but also congestion, and if it's caused by congestion you really *don't* want to increase the buffer sizes. That would just make the problem worse.

Reply:
There are two problems with implementing (1):

1. `utp_socket_impl` does not have access to neither `m_sm->m_sett`, nor `m_sm->m_sock` - both are declared private in `utp_socket_manager.hpp`. These are the only sources of information about `recv_socket_buffer_size`. Everything is so well isolated in that code... Breaking this isolation would make it ugly.

2. If we increase `m_in_buf_size` on increasing `recv_socket_buffer_size`, we increase it for **every** uTP stream out there, regardless of the real necessity to do that. This is clearly an overkill. We only want it increased on high-latency connections.

Therefore, I advocate to reconsider option (3). We could do that on basis of current latency data in LEDBAT.

Reply:
iirc, some settings are exposed via the utp socket manager (via function calls). So this could be exposed similarly. I agree with (2), it would be ideal to set it adaptively. I would like to make three points though:

1. The receive buffer size (which I suppose may not work right for uTP sockets right now) already has the problem you describe in (2). i.e. it would not be new. It affects TCP sockets the same way for instance.

2. The delay measurements from LEDBAT are only *relative* *one-way* delays. You can't estimate the bandwidth delay product from them. You could use RTT though, which is also maintained (as a running/sliding average). The problem there is still that the total delay you experience on a line may just as likely be due to congestion as to speed-of-light. I think this is fundamentally a difficult problem and a bit risky to mess too much with. The best solution I can think of is to track the *minimum* RTT ever experienced on a connection, and use that for the "physical" latency.

3. I think it's clear that the reorder count and the receive buffer size are intimately coupled. They both represent the number of incoming packet's we're willing to hold while waiting, either for the upper layer to consume more payload, or a lost packet to be re-sent. If the receive buffer (and the number of "reorder packets") adapts to network conditions, it's critical to ensure an adversarial counterpart can't DoS the client.

Reply:
Added method `adjust_receive_buf_size`. It adjusts `m_in_buf_size` linearly by adding 1Mbyte to it for each 100ms RTT latency. It would always clamp `1Mbyte<m_in_buf_size<8Mbytes`.
Testing show it works for high and low latencies as intended.

Reply:
However, during testing I noticed that `min_rtt` is indeed a local variable, and could not only shrink, but actually grow. And logs show it does, regularly. Is it by design? I thought it was for keeping the smallest seen latency during connection with a single peer (that is a single uTP stream, right?).

Besides, we have another serious problem: when I add some _variation_ to latency, both patched and unpatched versions become very slow, even for moderate latencies, like 200ms RTT. I will investigate this issue further.

Reply:
yeah, the current use for min_rtt is as a sanity check for the one-way delay measurements. As a protection against a malicious end-point, the one way delay measurement is clamped to the minimum RTT of any of the packets that were just ACKed. To keep a per-connection min_rtt you would need to add a new member (e.g. ``m_min_rtt``).

As for the issue you see with jitter, this is most likely because the one-way delay measurements are used to infer congestion. If you have non-congestion jitter, uTP will still interpret the high delay measurements as a signal to slow down, and the low delays as a signal to speed up.

Since you may expect jitter, you could increase the target delay. That's a session-wide setting. It defaults to 100ms, if you expect to have a range of [0,200] ms additional jitter, you should raise it by 200. Keep in mind the one-way delay measurement does not take into account any fixed latency. Like switch-processing or speed-of-light. It *only* looks at the variation in delays.